src/compiler/checker.ts

```diff
@@ -6011,7 +6017,88 @@ namespace ts {
+                                            // This is to deal with cases like below:
+                                            //
+                                            //     let x: A | B | C;
+                                            //
+                                            //     if (isB(x)) { // narrow x to A.
+                                            //     }
+                                            //     else if (isB(x)) { // narrow x to A
+                                            //     }
+                                            //     else if (isC(x)) { // narrow x to A | B
+                                            //     }
+                                            //     else { // x begins with A | B | C
+                                            //         x
+                                            //     }
+                                            //
+                                            // Which is the same as:
+                                            //
+                                            //     let x: A | B | C;
+                                            //
+                                            //     if (isA(x)) { // narrow x to A (but also removes A).
+                                            //     }
+                                            //     else if (isB(x)) { // narrow x to A
+                                            //     }
+                                            //     else if (isC(x)) {// narrow x to A | B
+                                            //     }
+                                            //     else { //x begins with A | B | C
+                                            //         x
+                                            //     }
+                                            //
+                                            // But we want only the latter to narrow the type to an empty object type.
+                                            //
+                                            type === narrowType(originalType, (<IfStatement>node).expression, /*assumeTrue*/ true) &&
+                                            // If the first narrowed type yield a distinct type(not a union type). And it encounters
+                                            // a narrowed type that yields the same type in some upper branches. It would then narrow
+                                            // the type to an empty object type instead of the distinct type. The below check, guards
+                                            // us from narrowing it to an empty object type instead of the distinct type.
+                                            //
+                                            //     if (isA(x)) { // narrow x to an empty object type
+                                            //     }
+                                            //     else if (isB(x)) { // narrow x to A

```

Comment:
The first else if clause below narrow the union type to A. This else if clause remove B from A.